lib/ui/painting.dart

```diff
@@ -1606,14 +1618,150 @@ class Image extends NativeFieldWrapperClass2 {
   /// Returns an error message on failure, null on success.
   String? _toByteData(int format, _Callback<Uint8List?> callback) native 'Image_toByteData';
 
+  bool _disposed = false;
   /// Release the resources used by this object. The object is no longer usable
   /// after this method is called.
-  void dispose() native 'Image_dispose';
+  ///
+  /// All outstanding handles from [createHandle] should be disposed before
+  /// calling this. Disposing all outstanding handles will automatically
+  /// dispose this object.
+  void dispose() {
+    assert(() {
+      assert(!_disposed);
+      assert(
+        _handles.isEmpty,
+        'Attempted to dispose of an Image object that has ${_handles.length} open handles.',
+      );
+      _disposed = true;
+      return true;
+    }());
+    _dispose();
+
+  }
+
+  void _dispose() native 'Image_dispose';
+
+  /// Returns the native wrapper of the Image. Any calls to `native` must use
+  /// this getter to avoid passing an [_ImageHandle] to native, which will
+  /// crash.
+  Image get _unwrapped => this;
+
+  Set<_ImageHandle> _handles = <_ImageHandle>{};
+
+  /// If asserts are enabled, returns the [StackTrace]s of each open handle from
+  /// [createHandle], in creation order.
+  ///
+  /// If asserts are disabled, this method always returns null.
+  List<StackTrace>? debugGetOpenHandleStackTraces() {
+    assert(!_disposed);
+    List<StackTrace>? stacks;
+    assert(() {
+      stacks = _handles.map((_ImageHandle handle) => handle.debugStack!).toList();
+      return true;
+    }());
+    return stacks;
+  }
+
+  /// Creates a disposable handle to this image.
+  ///
+  /// The returned object behaves identically to this image, except calling
+  /// [dispose] on it will only dispose the underlying native resources if it
+  /// is the last remaining handle.
+  Image createHandle() {
+    if (_disposed) {
+      throw StateError('Object disposed');
```

Comment:
can we come up with a better error message?

Reply:
I believe this is what is used in other cases when a native object is used that is disposed, but I'm open to suggestions.

Reply:
Something like "Cannot clone a disposed image.\nThe clone() method of a previously-disposed Image was called. Once an Image object has been disposed, it can no longer be used to create handles, as the underlying data may have been released."

Reply:
Done